components/edit.js

```diff
@@ -0,0 +1,235 @@
+  const router = useRouter()
+
+  useEffect(() => {
+    const isEdit = editing(document.URL)
+    // If the URL says edit but we aren't editing yet, set the state
+    if (isEdit && !edit) {
+      setEditing(true)
+    }
+
+    // If the URL has us not editing but we just were, set to false, and update props
+    if (!isEdit && edit) {
+      setEditing( false )
+      router.replace(router.asPath)
+    }
+  }, [edit, router])
+
+  const editpage = <EditPage item={item}/>
+  const editlink = <EditLink item={item}/>
+
+  const componentToShow = edit ? editpage :
+  <>
+    {viewComponent}
+    {editlink}
+  </>
+  return componentToShow
+}
+
+export const EditableItem = ({ item, children }) => {
+  return useEditor(item, children)
+}
+
+EditableItem.propTypes = {
+  item: PropTypes.object.isRequired,
+}
+
+const SaveCancelControl = ({
+  cancelClick,
+  saveClick,
+  itemPath,
+  saveEnabled,
+}) => {
+  return (
+    <div className="flex space-x-1">
+      <Button.Link href={itemPath} navigationClick={cancelClick}>
+        Cancel
+      </Button.Link>
+      <Button
+        onClick={saveClick}
+        enabled={saveEnabled}
+        type={saveEnabled ? "primary" : "info"}
+      >
+        Save
+      </Button>
+    </div>
+  )
+}
+
+SaveCancelControl.propTypes = {
+  cancelClick: PropTypes.func.isRequired,
+  saveClick: PropTypes.func.isRequired,
+  itemPath: PropTypes.string.isRequired,
+  saveEnabled: PropTypes.bool.isRequired,
+}
+
+function sortedJson(obj) {
+  if (Array.isArray(obj)) {
+    return obj.map((value) => sortedJson(value))
+  }
+  // We know it's not an array if we're here because the above `if`
+  if (typeof obj == "object") {
+    const sorted = {}
+    Object.keys(obj)
+      .sort()
+      .forEach((key) => {
+        sorted[key] = obj[key]
+      })
+    return sorted
+  }
+  return obj
+}
+
+const SavedErrors = ({ errors = [] }) => {
+  return (
+    <div>
+      <p className="text-lg text-rose-600">Errors from Save:</p>
+      <ul className="list-disc">
+        {errors.map((error) => (
+          <li key={error.description} className="text-base text-rose-600">
+            {error.description}
+          </li>
+        ))}
+      </ul>
+    </div>
+  )
+}
+
+SavedErrors.propTypes = {
+  errors: PropTypes.array.isRequired,
+}
+
+const EditPage = ({ item }) => {
+  const path = item["@id"]
+
+  const loggedIn = useAuthenticated()
+
+  const editable = (session, item) => {
+    // cannot edit if not logged in or object not editable
+    const editable = canEdit(item)
+    return loggedIn && editable
+  }
+
+  const jsonErrors = (json) => {
+    // cannot save if we cannot edit or if the JSON is wrong
+    try {
+      JSON.parse(json)
+      return []
+    } catch (err) {
+      // Save the error
+      return [err.message]
+    }
+  }
+
+  const { session } = useContext(SessionContext)
+
+  const [text, setText] = useState(() => {
+    JSON.stringify({}, null, 4)
+  })
+
+  const [saveErrors, setErrors] = useState([])
```

Comment:
Unusual naming for a state. Usually you see `[errors, setErrors]`, like you have below with `editorStatus`.

Reply:
what about `setSaveErrors`? Just trying to indicate that the errors are from a "save".

Reply:
I think it’d be less confusing for future readers to just comment that the errors come from a save, since this is so unusual for state naming. I have a personal policy of commenting 100% of my states, instead of just those that I deem unclear at that moment, when they’ll probably _all_ be unclear in a few months or less.

Reply:
Also, I learned from Keenan not to be _too_ shy about long-ass explanatory variable names. I try to avoid it just because it’s kind of a pain, but Keenan’s Javascript is very clear partly for that reason.